packages/@ember/-internals/glimmer/lib/utils/weak.ts

```diff
@@ -0,0 +1,86 @@
+import { Dict } from '@glimmer/interfaces';
+import { dict } from '@glimmer/util';
+
+const REFS = new WeakMap<object, WeakRef>();
+const VALS = new WeakMap<WeakRef, object>();
+
+export class WeakRef<T extends object = object> {
+  static for<T extends object>(value: T | WeakRef<T>): WeakRef<T> {
+    if (value instanceof WeakRef) {
+      return value;
+    }
+
+    let ref = REFS.get(value) as WeakRef<T> | undefined;
+
+    if (ref === undefined) {
+      ref = new WeakRef();
+      VALS.set(ref, value);
+      REFS.set(value, ref);
+    }
+
+    return ref;
+  }
+
+  static release<T extends object>(value: T | WeakRef<T>): void {
+    let ref = WeakRef.for(value);
+    let val = VALS.get(ref);
+
+    VALS.delete(ref);
+
+    if (val !== undefined) {
+      REFS.delete(val);
+    }
+  }
+
+  private constructor() {}
+
+  get(): T | undefined {
+    return VALS.get(this) as T | undefined;
+  }
+}
+
+const SETS = new WeakMap<object, Dict<WeakRefSet>>();
+
+export class WeakRefSet<T extends object = object> {
```

Comment:
Same here RE: inline docs, it is **very** hard to understand the broad thing that this is doing...